tls/extensions/s2n_server_key_share.c

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/extensions/s2n_server_key_share.h"
+
+#include "tls/s2n_client_extensions.h"
+#include "utils/s2n_safety.h"
+
+/*
+ * Calculate the data length for Server Key Share extension
+ *
+ * negotiated_curve should be selected in server_ecc_params.
+ */
+int s2n_extensions_server_key_share_send_size(struct s2n_connection *conn)
+{
+    const struct s2n_ecc_named_curve* curve = conn->secure.server_ecc_params.negotiated_curve;
+
+    if (curve == NULL) {
+        return 0;
+    }
+
+    const int key_share_size = S2N_SIZE_OF_EXTENSION_TYPE
+        + S2N_SIZE_OF_EXTENSION_DATA_SIZE
+        + S2N_SIZE_OF_NAMED_GROUP
+        + S2N_SIZE_OF_KEY_SHARE_SIZE
+        + curve->share_size;
+
+    return key_share_size;
+}
+
+/*
+ * Sends Key Share extension in Server Hello.
+ *
+ * Expects negotiated_curve to be set and generates a ephemeral key for key sharing
+ */
+int s2n_extensions_server_key_share_send(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    notnull_check(conn->secure.server_ecc_params.negotiated_curve);
+    notnull_check(out);
+
+    GUARD(s2n_stuffer_write_uint16(out, TLS_EXTENSION_KEY_SHARE));
+    GUARD(s2n_stuffer_write_uint16(out, s2n_extensions_server_key_share_send_size(conn) - S2N_SIZE_OF_EXTENSION_TYPE));
+
+    GUARD(s2n_ecdhe_parameters_send(&conn->secure.server_ecc_params, out));
+
+    return 0;
+}
+
+/*
+ * Client receives a Server Hello key share.
+ *
+ * If the curve is supported, conn->secure.server_ecc_params will be set.
+ */
+int s2n_extensions_server_key_share_recv(struct s2n_connection *conn, struct s2n_stuffer *extension)
+{
+    notnull_check(conn);
+    notnull_check(extension);
+
```

Comment:
We need a check here to skip processing this extension if we aren't possibly negotiating 1.3. Similarly for processing the other new extensions.

It seems the getenv() approach is the best until we have a new config preference to turn on 1.3.

Reply:
let me think if there's a better way to do this, since I'm not too sure about the cost of `getenv()`. i would prefer `#ifdef`s for all tls1.3 related stuff so it would optimized during compile time, but we can also skip parsing the extension when tls 1.3 server is negotiating with 1.2.

Reply:
I think we can go with a global variable to avoid processing 1.3 extensions. I've added a new issue to track this. https://github.com/awslabs/s2n/issues/1158

Reply:
Sure, that sounds good. Let's wait on that one before we merge this.